Use const generics for stdarch intrinsics

# Background

https://github.com/rust-lang/stdarch/issues/248

`core::arch` contains architecture-specific vendor intrinsics, some of which have arguments that are required to be compile-time constants. This is tricky because LLVM will often assert/ICE if a constant is not passed into the LLVM intrinsic at the LLVM IR level. This was previously handled by using a `match` statement to manually monomorphize the intrinsic:

```rust
#[inline]
#[target_feature(enable = "sse")]
#[cfg_attr(test, assert_instr(shufps, mask = 3))]
#[rustc_args_required_const(2)]
#[stable(feature = "simd_x86", since = "1.27.0")]
pub unsafe fn _mm_shuffle_ps(a: __m128, b: __m128, mask: i32) -> __m128 {
    let mask = (mask & 0xFF) as u8;

    macro_rules! shuffle_done {
        ($x01:expr, $x23:expr, $x45:expr, $x67:expr) => {
            simd_shuffle4(a, b, [$x01, $x23, $x45, $x67])
        };
    }
    macro_rules! shuffle_x67 {
        ($x01:expr, $x23:expr, $x45:expr) => {
            match (mask >> 6) & 0b11 {
                0b00 => shuffle_done!($x01, $x23, $x45, 4),
                0b01 => shuffle_done!($x01, $x23, $x45, 5),
                0b10 => shuffle_done!($x01, $x23, $x45, 6),
                _ => shuffle_done!($x01, $x23, $x45, 7),
            }
        };
    }
    macro_rules! shuffle_x45 {
        ($x01:expr, $x23:expr) => {
            match (mask >> 4) & 0b11 {
                0b00 => shuffle_x67!($x01, $x23, 4),
                0b01 => shuffle_x67!($x01, $x23, 5),
                0b10 => shuffle_x67!($x01, $x23, 6),
                _ => shuffle_x67!($x01, $x23, 7),
            }
        };
    }
    macro_rules! shuffle_x23 {
        ($x01:expr) => {
            match (mask >> 2) & 0b11 {
                0b00 => shuffle_x45!($x01, 0),
                0b01 => shuffle_x45!($x01, 1),
                0b10 => shuffle_x45!($x01, 2),
                _ => shuffle_x45!($x01, 3),
            }
        };
    }
    match mask & 0b11 {
        0b00 => shuffle_x23!(0),
        0b01 => shuffle_x23!(1),
        0b10 => shuffle_x23!(2),
        _ => shuffle_x23!(3),
    }
}
```

The downside of this approach is that it generates huge MIR and severely bloats the size of `libcore` to the point where it actually affects compilation time.

The `#[rustc_args_required_const]` attribute was used to restrict arguments to constants so that we could eventually replace this with const generics. Note that it effectively only acts as a hard lint, the `match` is still needed to manually monomorphize the LLVM intrinsic.

# New const generics support

https://github.com/rust-lang/stdarch/issues/1022

All the intrinsics in `stdarch` have been converted to use the newly stabilized const generics:

```rust
#[inline]
#[target_feature(enable = "sse")]
#[cfg_attr(test, assert_instr(shufps, mask = 3))]
#[rustc_legacy_const_generics(2)]
#[stable(feature = "simd_x86", since = "1.27.0")]
pub unsafe fn _mm_shuffle_ps<const mask: i32>(a: __m128, b: __m128) -> __m128 {
    static_assert!(mask: i32 where mask >= 0 && mask <= 255);
    simd_shuffle4(
        a,
        b,
        [
            mask as u32 & 0b11,
            (mask as u32 >> 2) & 0b11,
            ((mask as u32 >> 4) & 0b11) + 4,
            ((mask as u32 >> 6) & 0b11) + 4,
        ],
    )
}
```

To preserve backwards compatibility with the already stabilized intrinsics using `#[rustc_args_required_const]`, a new attribute `#[rustc_legacy_const_generics]` was added in #82447 which rewrites function calls of the form `func(a, b, c)` to `func::<{b}>(a, c)`.

This new attribute is not intended to ever be stabilized, it is only intended for use in `stdarch` as a replacement for `#[rustc_args_required_const]`.

```rust
#[rustc_legacy_const_generics(1)]
pub fn foo<const Y: usize>(x: usize, z: usize) -> [usize; 3] {
    [x, Y, z]
}

fn main() {
    assert_eq!(foo(0 + 0, 1 + 1, 2 + 2), [0, 2, 4]);
    assert_eq!(foo::<{1 + 1}>(0 + 0, 2 + 2), [0, 2, 4]);
}
```

# Open questions/issues

- `#[rustc_args_required_const]` allowed constant expressions derived from generic arguments in the caller, but this is not allowed with const generics. This technically makes this a breaking change.
- Another potential breaking change is that the intrinsics now use a post-monomorphization `static_assert` to validate constant arguments. Previously, intrinsics would either use runtime asserts or simply pick an arbitrary default behavior for out-of-range constants.
- Do we want to force all callers to use the const generics syntax in the 2021 edition?
  - Pros:
    - It makes it clearer that these intrinsics expect a constant as an argument.
    - Intrinsics are rendered with the const generics syntax in rustdoc.
  - Cons:
    - We will still need to support the `#[rustc_legacy_const_generics]` argument rewriting anyways for 2015/2018 code.
    - Intrinsics no longer have the same function signature as the original C vendor intrinsics.

So there are a couple sources of breaking changes in the `stdarch` bump that have been hit in the crater run:
- one is related to using associated consts of generic parameters
- the other one is that a few shift-related AVX2 intrinsics were stabilized without the `#[rustc_args_required_const]` attribute that ensures the immediates are constants. Two of those were hit in the crater run, and I'll explain those below.

https://github.com/rust-lang/stdarch/pull/1025/files#diff-95541a10736e81b6d3dd51a3a94005c39f9ebc49fdf44c2956f925f74c39183a is Amanieu's PR noticing and fixing this by updating these intrinsics to const generics: they added the `#[rustc_legacy_const_generics]` attribute which for all intents and purposes in this explanation is the same as if they had only added the missing `#[rustc_args_required_const]` attribute.

[Here's an example](https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=b2e1e20deb99afba21071095823a37be) of the difference on stable: the AVX2 example compiles even though the immediate is not a constant, while the SSE2 version only compiles when passed a constant immediate. (Amanieu explained above why it can also run fine).

The crater cases were very similar to these examples:
- different versions of `simdeez`: [using `_mm256_srli_epi32`](https://github.com/jackmott/simdeez/blob/796ac4d7817f066b5ccc2a7b11fdca67d83da192/src/avx2/avx2.rs#L530-L533) which caused [this error](https://crater-reports.s3.amazonaws.com/pr-83278/try%239cae4dca0b4ddbfe82616f4d5f1090fc817e2ffa/reg/simdeez-1.0.6/log.txt) for example on simdeez-1.0.6
```
error[E0435]: attempt to use a non-constant value in a constant
[INFO] [stdout]    --> src/avx2/avx2.rs:527:38
[INFO] [stdout]     |
[INFO] [stdout] 521 |               let mask = _mm256_setr_epi32(0, -1, 0, -1, 0, -1, 0, -1); // mask for high part containing only sign
[INFO] [stdout]     |  ______________-
[INFO] [stdout] 522 | |             Self::blendv_epi64(I64x4(sra3), I64x4(sign), I64x4(mask))
[INFO] [stdout] 523 | |         }
[INFO] [stdout] 524 | |     }
[INFO] [stdout] 525 | |     #[inline(always)]
[INFO] [stdout] 526 | |     unsafe fn srli_epi32(a: Self::Vi32, amt_const: i32) -> Self::Vi32 {
[INFO] [stdout]     | |_________________________________________________- help: consider using `const` instead of `let`: `const amt_const`
[INFO] [stdout] 527 |           I32x8(_mm256_srli_epi32(a.0, amt_const))
[INFO] [stdout]     |                                        ^^^^^^^^^ non-constant value
```
- `highway-0.6.3`: [using `_mm256_srli_epi64`](https://github.com/nickbabcock/highway-rs/blob/68ce47adaa4fd5a0b0c3cc9db8994254e1e2b1e8/src/v4x64u.rs#L170-L178) in the `Shr` impl of their custom vector type and which caused [this error](https://crater-reports.s3.amazonaws.com/pr-83278/try%239cae4dca0b4ddbfe82616f4d5f1090fc817e2ffa/reg/highway-0.6.3/log.txt)
```
error[E0435]: attempt to use a non-constant value in a constant
[INFO] [stdout]    --> src/v4x64u.rs:175:52
[INFO] [stdout]     |
[INFO] [stdout] 164 |           let mut new = V4x64U(self.0);
[INFO] [stdout]     |  __________-
[INFO] [stdout] 165 | |         new ^= other;
[INFO] [stdout] 166 | |         new
[INFO] [stdout] 167 | |     }
[INFO] [stdout] ...   |
[INFO] [stdout] 172 | |
[INFO] [stdout] 173 | |     fn shr(self, shift: i32) -> Self {
[INFO] [stdout]     | |______________________- help: consider using `const` instead of `let`: `const shift`
[INFO] [stdout] 174 |           let mut new = V4x64U(self.0);
[INFO] [stdout] 175 |           unsafe { new.0 = _mm256_srli_epi64(self.0, shift) }
[INFO] [stdout]     |                                                      ^^^^^ non-constant value
``` 

`simdeez` also uses the same technique as `stdarch` used to (before the switch to const generics) to ensure the immediates are effectively constant: a [macro covering all the values the immediate can take](https://github.com/jackmott/simdeez/blob/796ac4d7817f066b5ccc2a7b11fdca67d83da192/src/macros.rs#L9) and so [their fix](https://github.com/jackmott/simdeez/commit/2bd23df3068512172f44a663c6939440d6563139#diff-24d7b54d4b18942a95306088e6c4cd6e1e90267d616feb9a5b3b0629f9966ad4) was to use this for the intrinsic that should have required it.

`highway` does not use such a heavy trick (and funnily enough the author had noticed the discrepancy between their SSE and AVX code implementations, only the AVX code path could have that `Shr` impl). Since their only use of this custom operator was [with a constant shift amount](https://github.com/nickbabcock/highway-rs/blob/68ce47adaa4fd5a0b0c3cc9db8994254e1e2b1e8/src/avx.rs#L214-L222) it was easy to remove. 

I've seen LLVM excel in optimizing similar mathematical expressions with constants before, even with custom operators like `highway` did, and that could explain why the crates work fine today. But it seems that even if that's not the case, Amanieu's investigation shows that at least some of these intrinsics were made to handle runtime values instead of only constant immediates as one would expect from the official docs.

--- a/src/librustdoc/clean/mod.rs
+++ b/src/librustdoc/clean/mod.rs
@@ -749,7 +749,38 @@ fn clean_fn_or_proc_macro(
                 } else {
                     hir::Constness::NotConst
                 };
+            clean_fn_decl_legacy_const_generics(&mut func, attrs);
             FunctionItem(func)
+        }
+    }
+}
+
+/// This is needed to make it more "readable" when documenting functions using
+/// `rustc_legacy_const_generics`. More information in
+/// <https://github.com/rust-lang/rust/issues/83167>.
+fn clean_fn_decl_legacy_const_generics(func: &mut Function, attrs: &[ast::Attribute]) {
+    for meta_item_list in attrs
+        .iter()
+        .filter(|a| a.has_name(sym::rustc_legacy_const_generics))
+        .filter_map(|a| a.meta_item_list())
+    {
+        for (pos, literal) in meta_item_list.iter().filter_map(|meta| meta.literal()).enumerate() {
+            match literal.kind {
+                ast::LitKind::Int(a, _) => {
+                    let gen = func.generics.params.remove(0);
+                    if let GenericParamDef { name, kind: GenericParamDefKind::Const { ty, .. } } =
+                        gen
+                    {
+                        func.decl
+                            .inputs
+                            .values
+                            .insert(a as _, Argument { name, type_: *ty, is_const: true });
+                    } else {
+                        panic!("unexpected non const in position {}", pos);
+                    }
+                }
+                _ => panic!("invalid arg index"),
+            }
         }
     }
 }
@@ -779,7 +810,7 @@ impl<'a> Clean<Arguments> for (&'a [hir::Ty<'a>], &'a [Ident]) {
                     if name.is_empty() {
                         name = kw::Underscore;
                     }
-                    Argument { name, type_: ty.clean(cx) }
+                    Argument { name, type_: ty.clean(cx), is_const: false }
                 })
                 .collect(),
         }
@@ -798,6 +829,7 @@ impl<'a> Clean<Arguments> for (&'a [hir::Ty<'a>], hir::BodyId) {
                 .map(|(i, ty)| Argument {
                     name: name_from_pat(body.params[i].pat),
                     type_: ty.clean(cx),
+                    is_const: false,
                 })
                 .collect(),
         }
@@ -828,6 +860,7 @@ impl<'tcx> Clean<FnDecl> for (DefId, ty::PolyFnSig<'tcx>) {
                     .map(|t| Argument {
                         type_: t.clean(cx),
                         name: names.next().map_or(kw::Empty, |i| i.name),
+                        is_const: false,
                     })
                     .collect(),
             },

--- a/src/librustdoc/clean/types.rs
+++ b/src/librustdoc/clean/types.rs
@@ -1353,6 +1353,9 @@ crate struct Arguments {
 crate struct Argument {
     crate type_: Type,
     crate name: Symbol,
+    /// This field is used to represent "const" arguments from the `rustc_legacy_const_generics`
+    /// feature. More information in <https://github.com/rust-lang/rust/issues/83167>.
+    crate is_const: bool,
 }
 
 #[derive(Clone, PartialEq, Debug)]

--- a/src/librustdoc/html/format.rs
+++ b/src/librustdoc/html/format.rs
@@ -1177,6 +1177,10 @@ impl clean::FnDecl {
                     args.push_str(" <br>");
                     args_plain.push(' ');
                 }
+                if input.is_const {
+                    args.push_str("const ");
+                    args_plain.push_str("const ");
+                }
                 if !input.name.is_empty() {
                     args.push_str(&format!("{}: ", input.name));
                     args_plain.push_str(&format!("{}: ", input.name));

--- a/src/test/rustdoc/legacy-const-generic.rs
+++ b/src/test/rustdoc/legacy-const-generic.rs
@@ -1,0 +1,16 @@
+#![crate_name = "foo"]
+#![feature(rustc_attrs)]
+
+// @has 'foo/fn.foo.html'
+// @has - '//*[@class="rust fn"]' 'fn foo(x: usize, const Y: usize, z: usize) -> [usize; 3]'
+#[rustc_legacy_const_generics(1)]
+pub fn foo<const Y: usize>(x: usize, z: usize) -> [usize; 3] {
+    [x, Y, z]
+}
+
+// @has 'foo/fn.bar.html'
+// @has - '//*[@class="rust fn"]' 'fn bar(x: usize, const Y: usize, const Z: usize) -> [usize; 3]'
+#[rustc_legacy_const_generics(1, 2)]
+pub fn bar<const Y: usize, const Z: usize>(x: usize) -> [usize; 3] {
+    [x, Y, z]
+}

